### PR TITLE
Allow checking subarray corner positions and V2/V3 reference points 

### PR DESCRIPTION
**Pull request by mgennaro**
NIRCam wanted to be able to explicitly verify the consistency of SIAF apertures definitions w.r.t. the subarray corners that available in the subarray file in the PRD.

To do this I created a '''DMS_corner''' method of the aperture class that allows computing the lower-left corner position in the OSS/DMS coordinate system (for which there is a 0.5 offset w.r.t. to the siaf definitions). 

I also modified the generate_nircam script to produce an "enhnaced aperture file" that contains both the ColCorner/RowCorner entries (the OSS name for the subarray lower-left corner coordinates) as well as the V2/V3 coordinates of the aperture reference point. 
Having the explicit values of V2/V3 will allow NIRCam to verify that the V2/V3 values used by OSS (e.g. for TA) are consistent with what expected according to SIAF. 
The enhanced aperture file is created by running generate_nircam and it's placed in the pre_delivery_data directory. The "official" siaf aperture definition file in untouched by this process.

In the NIRCam team SIAF delivery process, the new enhanced file will allow further cross-checking of any changes we might introduce, in addition to the cross-check already provided by the diagnostic plots.

@JAStansberry @marthaboyer

### CODE REVIEW

File: pysiaf/aperture.py [hunk 1]
```diff
@@ -1437,8 +1437,17 @@ def verify(self):
                 print('Verification WARNING: {} has non-integer sum between YDetRef {} and '
                       'YSciRef {} DetSciYAngle {}'.format(self.AperName, self.YDetRef, self.YSciRef,
                                                           self.DetSciYAngle))
-
-
+    def DMS_corner(self):
```
Comment (Johannes-Sahlmann) [inline pysiaf/aperture.py]: Method names should be all lower case.

File: generate/generate_nircam.py [hunk 2]
```diff
@@ -441,3 +405,94 @@
...
+
+in_file = os.path.join(JWST_SOURCE_DATA_ROOT,instrument,'nircam_siaf_aperture_definition.txt')
+out_file = os.path.join(JWST_DELIVERY_DATA_ROOT, instrument,'nircam_enhanced_siaf_aperture_definition.txt')
+
+print('Writing SIAF aperture definition file including subarrays and V2/V3 of reference point')
+with open(in_file) as fp:
+    lines = fp.readlines()
+
+
+with open(out_file, 'w') as the_file:
```
Comment (Johannes-Sahlmann) [inline generate/generate_nircam.py]: If what I mention in my main comment is true, this code will only be executed once and therefore does need to be part of the repo.

File: pysiaf/aperture.py (hunk 1 above)
Comment (mgennaro) [reply pysiaf/aperture.py]: Done

File: generate/generate_nircam.py (hunk 2 above)
Comment (mgennaro) [reply generate/generate_nircam.py]: We decided not to alter the current format of the nircam_siaf_aperture_definition.txt file, to avoid compatibility issues (also with other subsystems).
We would like the "enhanced" file to be produced every time a change needs to be introduced to nircam apertures. In this respect, the enhanced file has the same function as the "JIRA" plots you are already producing in generate_nircam. Please let me know if you'd like this piece of code to be moved to a different location within generate_nircam

File: generate/generate_nircam.py [hunk 3]
```diff
@@ -441,3 +405,94 @@
...
+
+
+
+
+
+
+
+
+
+
```
[the remark this answers is not in the document]
Comment (mgennaro) [reply generate/generate_nircam.py]: Done

File: generate/generate_nircam.py [hunk 4]
```diff
@@ -441,3 +405,94 @@
...
+                       
+                except:
+                    llist.insert(10,' {:>20} '.format('None'))
+                    llist.insert(11,' {:>20} '.format('None'))
+                else:            
+                    llist.insert(10,' {:20.14f} '.format(V2_ref))
+                    llist.insert(11,' {:20.14f} '.format(V3_ref))
+
+       
+
```
[the remark this answers is not in the document]
Comment (mgennaro) [reply generate/generate_nircam.py]: Done

File: generate/generate_nircam.py [hunk 5]
```diff
@@ -441,3 +405,94 @@
...
+             'NRCA5_FSTAMASK430R','NRCA4_TAMASKSWB','NRCA4_TAMASKSWBS','NRCA4_FSTAMASKSWB','NRCA5_TAMASKLWB','NRCA5_TAMASKLWBL','NRCA5_FSTAMASKLWB']
+
+in_file = os.path.join(JWST_SOURCE_DATA_ROOT,instrument,'nircam_siaf_aperture_definition.txt')
+out_file = os.path.join(JWST_DELIVERY_DATA_ROOT, instrument,'nircam_enhanced_siaf_aperture_definition.txt')
+
+print('Writing SIAF aperture definition file including subarrays and V2/V3 of reference point')
+with open(in_file) as fp:
+    lines = fp.readlines()
+
+
```
[the remark this answers is not in the document]
Comment (mgennaro) [reply generate/generate_nircam.py]: Done

File: generate/generate_nircam.py (hunk 2 above)
Comment (mgennaro) [reply generate/generate_nircam.py]: As for the ._* files...sorry will try not to '''git add *''' next time

Comment (Johannes-Sahlmann) [reply generate/generate_nircam.py]: The only customer of nircam_siaf_aperture_definition.txt in pysiaf, we we can add columns there without any problems.

Comment (mgennaro) [reply generate/generate_nircam.py]: Ok...let me rephrase that. The subarray corners, the way I defined this enhanced file, are output products, computed self-consistently in pysiaf. we want to use them to double check entries in the PRD subarray file, that's why we don't want to have them in input in the regular siaf aperture deifnition file....at least for now (we may want to have all info in a single file in the future, or at least we are discussing that). Moreover, although pysiaf might be the only customer, the nircam team has already shared the enhanced file with OSS for them to verify, e.g. that their v2/v3 ref positions in the TA scripts are consistent with what pysiaf computes. I think we really prefer having this enhnaced file be separated from the source file and avoid confusion.

Comment (Johannes-Sahlmann) [reply generate/generate_nircam.py]: Our current workflow implies that we purge the pre_delivery_data directory before we make a pysiaf release. Is this compatible with your use cases?

Comment (mgennaro) [reply generate/generate_nircam.py]: I believe so. Basically I'd like that file to be treated like the "JIRA plots": 

1. we can produce that file ourselves by running generate_nircam and looking at the pre_delivery products, to make sure that whatever we put in a pysiaf PR looks consistent to us
2. if that's ok with you, that file could be part of the package you guys ask us to verify before merging a PR (i.e. the one you attach to the JSOCINT issue)
3.  just like the aperture plots in that package, the file itself does not need to be included in a release

For these reasons I believe it may make more sense to move the piece of code that creates the enhanced file inside the ```if emulate_delivery``` statement.  My new proposal is:

- I could package those ~60 lines of code into a separate function within the pysiaf.utils module and call it ```create_enhanced_aperture_file```

- I could place an ```if enhanced_aperture_file:  create_enhanced_aperture_file```
 statement after the ```for compare_to in [pysiaf.JWST_PRD_VERSION]:``` loop and before the ```    # run some tests on the new SIAF``` one 

Let me know what you think

Comment (Johannes-Sahlmann) [reply generate/generate_nircam.py]: Thank you for clearing up our questions. 
I like your suggested implementation. 

@shanosborne , do you agree, too?

Comment (mgennaro) [reply generate/generate_nircam.py]: My new commit should address this last point

Comment (shanosborne) [reply generate/generate_nircam.py]: I like the change of putting the new code under the the `if emulate_delivery` statement, it clearly groups this new files with all the other analysis files. Looks good to me!

File: pysiaf/utils/enhanced_aperture_file.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+-------
+    Mario Gennaro
+
+"""
+
+#from __future__ import absolute_import, print_function, division
+import os
+from ..constants import JWST_SOURCE_DATA_ROOT, JWST_TEMPORARY_DATA_ROOT, JWST_DELIVERY_DATA_ROOT
+
+def create_enhanced_aperture_file(aperture_dict,verbose=False):
```
[the remark this answers is not in the document]
Comment (mgennaro) [reply pysiaf/utils/enhanced_aperture_file.py]: Done and Done

**Comment by Johannes-Sahlmann** [comment]
there are some ._ files left

File: pysiaf/aperture.py [hunk 7]
```diff
@@ -1437,8 +1437,17 @@ def verify(self):
...
                                                           self.DetSciYAngle))
-
-
+    def dms_corner(self):
+        """Compute the pixel value of the lower left corner of an aperture.
+        
+        This corresponds to the OSS corner position (x: ColCorner, y: RowCorner).
+        The notation for OSS is 1-based, i.e. the lower left corner of a FULL subarray is (1,1)
+        """
+        ColCorner = np.ceil(np.min(self.corners('det')[0])).astype(np.int_)
```
Comment (Johannes-Sahlmann) [inline pysiaf/aperture.py]: Adhere to PEP8 variable naming conventions: use `column_corner` and `row_corner` names.

Comment (mgennaro) [reply pysiaf/aperture.py]: done, and spotted and fixed the same in enhanced_aperture_file. Also fixed an other couple of instances (those I could spot) of non pep8 compliant variable names
